### ISSUE TITLE
How does transaction fee & weights plays with XCMP/HRMP?

### ISSUE THREAD

**Issue by pepyakin** [paritytech/polkadot#1669]
How does transaction fee & weights plays with this?
Also the dest parachain may choose to ignore some messages due to fee / weight requirement, and in that case how does the sender know the message is been dropped?

_Originally posted by @xlc in https://github.com/paritytech/polkadot/issues/1641#issuecomment-684070555_

**Comment by pepyakin** [comment]
Fees & weights [for the work taken on the dest parachain] are outside of the XCMP & HRMP mechanisms. 

The protocols should be designed in a way that guarantee that a well behaved sender's messages would fit into the weight requirement of the dest chain. (And the original [issue](github.com/paritytech/polkadot/issues/1641)) where this question was asked is meant to help to ensure this)

As to how would the dest chain get fees - I am not sure. But maybe there is no need in fees, at least for well behaved parachains. If the sender sends unexpectedly too much messages it could be throttled after all instead of dropping.

**Comment by pepyakin** [comment]
Closing since the question was addressed. Feel free to reopen if you have any further questions, @xlc !

**Comment by xlc** [comment]
I am not sure how do you define / enforce a parachain to become a well behaved sender.

As of current, the weight & fee on a chain only enforce the resource usage on the current chain. For a cross-chain message, there will be some resource consumption (computation & storage) on destination chain as well, which I don't think it is something the sending chain can cope with.

One basic example: Alice send 1 DOT from chain A to chain B, the time required to process this is 1ms on chain A and 1ms on chain B and use 32 bytes on-chain state on chain B (to make new account).
How should chain A charge this transaction? Based on 1ms weight or 2ms weight?
What happen to chain B's validator / collator? Should they get compensation on the 1ms processing time as well as the additional storage usage? 

I don't think we can just assume / require all the chains to be well behaved. It is up to the chain's user to initiate cross-chain messages and I don't think the current tools (weight) we have is enough to regulate the users to not spam the cross-chain messages.

Image a PoA parachain that allow whitelisted user send transaction for free. In think economic incentive is the only way to ensure a fair usage.

**Comment by xlc** [comment]
With the new channel (didn't look into it much), one way will be require both chains to create a channel to allow free messages.

So it is up to parachains to come out mutual agreements first before they can send messages to each other.

One easy example will be for every 1000 ACA staked, the staked chain will be able to consume 1000000 weights per block on Acala parachain. Other chains can also design their own terms before open / accept a channel.

**Comment by pepyakin** [comment]
First of all, I beg your pardon because I used some sloppy terminology. Let's back up and try to unpack the problem.

XCMP & HRMP only allow communication through channels. A parachain can message another parachain only if the sender has a channel opened to the recipient. Channels are unidirectional and, obviously, only sender can send a message into an opened channel. 

On top of that:

- both parties should agree on the channel opening. Ultimately, the recipient confirms the channel, thus it is in control of the channel parameters (see #1641), notably, the message size.
- both the recipient and sender can close the channel between them at any time (with some delay, but that's irrelevant)

Channels can house a number of messages until they are read. The recipient doesn't have to read all the messages all at once (although a parathread needs to read messages within the availability window time frame lest missing the preimages disappearing).

But then, since the recipient blocks essentially are limited in some way, there are questions:

1. how does the recipient prevent abuse from the sender and 
1. how does the sender ensure that its messages are actually processed,

In particular, the first things that we care about are the processing power and the size of the message (because those will go into PoV, PoV size is what limits the amount of information that can be accessed during the block processing). We do care about the storage size, but arguably not as much as the processing power and sizes. For some paras the storage might be near-0 cost (e.g. imagine a stateless chain where storing the preimages is the responsibility of the asset holders and dropping a preimage will only hurt the corresponding holder). To make the matters worse, consider that the concept of weight makes sense only with respect to a particular version of code with a particular operation at hand, implying that even the same operation can have different weight within the same chain after an upgrade (nevermind the possibility of making the extrinsic invalid under the worst circumstances).

All of this makes me doubtful that we could introduce a common system of measuring the resources which both sender and recipient can rely on, in part, due to sheer possibility of forms and shapes of parachains. There is just no single answer to these questions above. If something works for one pair of sender-receiver might not work for another.

As a result, the current plan is to not embed any fees or weights into on the messaging protocol level (i.e. XCMP, HRMP). 

Instead, this is up to the parachains themselves to figure out the appropriate forms of abuse prevention as well as prioritization of messages. After all, incoming messages compete with transactions and, as a chain author, you need to come up with mitigation strategies for those as well. 

Admittedly, transactions (signed extrinsics) are easier to deal with since they assume an existing account which signs those. In the cross-chain scenario it might be trickier.

You already discovered one possible arrangement with locking some funds. That would work if there is some trust involved between chains (or rather their governance systems) - after all, you can do minimal amount of damage and then lose all credibility from all the other who might not want to establish a connection with your chain or close existing ones.

Perhaps, there could be some arrangements with payment channels established? Perhaps. If it is useful enough we might want to incorporate those on the messaging protocol level.

The bottom line is that these kind of arrangements are on the higher-level, i.e. level of agreements, right now. It is a part of the sender-recipient protocol as things as the message formats, the channel limits and the other things that should be agreed upon before opening a channel.

Hopefully, I managed to answer your question now. If my answer convoluted - sorry about that. But feel free to ask follow-up questions if you have any!

**Comment by xlc** [comment]
Thanks for the detailed explanation. 

Sounds like my proposal (lock up token for quota) is something we should explorer more.

Any ideas when this may land on Rococo so we can plan when to start implement this idea?

**Comment by pepyakin** [comment]
I think there are no big blockers to commence the planning stage as of now. Hopefully, the code for HRMP should land end of month. XCMP is way further in future and is more likely to be end of year.
